### PR TITLE
Add binary expression parsing

### PR DESCRIPTION
**Pull request by io12**
I think this works but so far I only tested it on "2 + 2"

### CODE REVIEW

**Comment by Alfriadox** [comment]
This looks mostly good, however I think we should make some changes to how we refer to structures (for example not all Expressions are necessarily Binary Expressions.) and I want to make sure the tests actually checks the correctness of the produced structure, instead of just printing it out to the terminal.

File: wright/src/grammar/parsers/expression/expression_tests.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    let h = files.add("val", val.to_string());
+    (files, h)
+}
+
+#[test]
+fn bin_op() {
+    let (files, h) = setup("2 + 2");
+    let frag = Fragment::new(&files, h);
+    let expr = Expression::parse(frag).unwrap();
+    println!("{:#?}", expr);
```
Comment (Alfriadox) [inline wright/src/grammar/parsers/expression/expression_tests.rs]: We should check correctness rather than just printing here.

**Comment by io12** [comment]
Alright I can add asserts and more test cases.

**Comment by io12** [comment]
I added correctness checks but the code is a bit repetitive (especially if we want to add more test cases). Maybe we should implement `Eq` for the expression nodes but ignore `Fragment` in the comparison? That way parser test cases could be as easy as comparing the AST result to an expected AST.

**Comment by Alfriadox** [comment]
> I added correctness checks but the code is a bit repetitive (especially if we want to add more test cases). Maybe we should implement `Eq` for the expression nodes but ignore `Fragment` in the comparison? That way parser test cases could be as easy as comparing the AST result to an expected AST.

I was thinking about that, but I don't think implementing Eq or PartialEq is the answer, since that's public outside the tests and could confuse anyone who uses wright as a library as to the functionality.

Later we will implement functionality that evaluates AST nodes. Then we can test by evaluation.

**Comment by io12** [comment]
A private `assert_expr_eq()` function might work. Having it call `assert_eq!()` instead of returning bool would output more useful info on failure.

**Comment by Alfriadox** [comment]
This may need to be re-implemented because the other branch that just got merged used all of this and then undid some of it

**Pull request merged**